src/services/reader/web_sys.rs

```diff
@@ -0,0 +1,160 @@
+//! `web-sys` implementation for the reader service.
+
+use crate::callback::Callback;
+use crate::services::Task;
+use gloo::events::EventListener;
+use js_sys::Uint8Array;
+use std::cmp;
+use std::fmt;
+pub use web_sys::{Blob, File};
+use web_sys::{Event, FileReader};
+
+/// Struct that represents data of a file.
+#[derive(Clone, Debug)]
+pub struct FileData {
+    /// Name of loaded file.
+    pub name: String,
+    /// Content of loaded file.
+    pub content: Vec<u8>,
+}
+
+/// Struct that represents a chunk of a file.
+#[derive(Clone, Debug)]
+pub enum FileChunk {
+    /// Reading of chunks started. Equals **0%** progress.
+    Started {
+        /// Name of loaded file.
+        name: String,
+    },
+    /// The next data chunk that read. Also provides a progress value.
+    DataChunk {
+        /// The chunk of binary data.
+        data: Vec<u8>,
+        /// The progress value in interval: `0 < progress <= 1`.
+        progress: f32,
+    },
+    /// Reading of chunks finished. Equals **100%** progress.
+    Finished,
+}
+
+/// A reader service attached to a user context.
+#[derive(Default, Debug)]
+pub struct ReaderService {}
+
+impl ReaderService {
+    /// Creates a new service instance connected to `App` by provided `sender`.
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    /// Reads all bytes from a file and returns them with a callback.
+    pub fn read_file(&mut self, file: File, callback: Callback<FileData>) -> ReaderTask {
+        let file_reader = FileReader::new().unwrap();
+        let reader = file_reader.clone();
+        let name = file.name();
+        let callback = move |_event: &Event| {
+            let array = Uint8Array::new_with_byte_offset(
+                &reader
+                    .result()
+                    .expect("`FileReader` hasn't finished loading"),
+                0,
+            );
+            let data = FileData {
+                name: name.clone(),
+                content: array.to_vec(),
+            };
+            callback.emit(data);
+        };
+        let listener = Some(EventListener::new(&file_reader, "loadend", callback));
```

Comment:
Looks like https://docs.rs/gloo-events/0.1.1/gloo_events/struct.EventListener.html#method.once is better suited for this event since it will only be called once

Reply:
Done.